docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+##### Memory randomization
+
+The alternative is to enable memory randomization. Before every iteration, the harness is going to allocate random-size objects, keep them alive and re-run the setup that should allocate the actual memory.
+
+You can read more about it [here](https://github.com/dotnet/BenchmarkDotNet/pull/1587). It requires an understanding of what distribution is and how to read it. It's also out of scope of this document, but a book on statistics, such as [Pro .NET Benchmarking](https://aakinshin.net/prodotnetbenchmarking/) can help you get a very good understanding of the subject.
+
+No matter how you are going to benchmark your code, you need to keep in mind that **the larger the input, the more you can benefit from vectorization**. If your code uses small buffers, performance might even get worse.
+
+## Loops
+
+To work with inputs that are bigger than a single vector, you typically need to loop over the entire input. This should be split into two parts:
+
+* vectorized loop that operates on multiple values at a time
+* handling of the remainder
+
+Example: our input is a buffer of ten integers, assuming that `Vector128` is accelerated, we handle the first four values in the first loop iteration, the next four in the second iteration and then we stop, as only two are left. Depending on how we can handle the remainder, we distinguish two approaches.
+
+### Scalar remainder handling
+
+Imagine that we want to calculate the sum of all the numbers in given buffer. We definitely want to add every element just once, without repetitions. That is why in the first loop, we add four (128 bits / 32 bits) integers in one iteration. In the second loop, we handle the remaining values.
+
+
+```cs
+int Sum(Span<int> buffer)
+{
+    Debug.Assert(Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<int>.Count);
+
+    // The initial sum is zero, so we need a vector with all elements initialized to zero.
+    Vector128<int> sum = Vector128<int>.Zero;
+
+    // We need to obtain the reference to first value in the buffer, it's used later for loading vectors from memory.
+    ref int searchSpace = ref MemoryMarshal.GetReference(buffer);
+    // And an offset, that is going to be used by vectorized and scalar loops.
+    nuint elementOffset = 0;
+    // And the last valid offset from which we can load the values
+    nuint oneVectorAwayFromEnd = (nuint)(buffer.Length - Vector128<int>.Count);
+    for (; elementOffset <= oneVectorAwayFromEnd; elementOffset += (nuint)Vector128<int>.Count)
+    {
+        // We load a vector from given offset.
+        Vector128<int> loaded = Vector128.LoadUnsafe(ref searchSpace, elementOffset);
+        // We add 4 integers at a time:
+        sum += loaded;
+    }
+
+    // We sum all 4 integers from the vector to one
+    int result = Vector128.Sum(sum);
+
+    // And handle the remaining elements, in a non-vectorized way:
+    while (elementOffset < (nuint)buffer.Length)
```

Comment:
The alternative here (and what is typically better for performance) is to backtrack, load one more vectors worth of data, and mask off any elements already processed.

This may be out of scope for the intent of the document at this point, but it might be worth calling out there are alternatives and possibly linking to an "advanced" section that covers such things.